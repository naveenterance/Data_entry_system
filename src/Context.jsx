import React, { createContext, useContext, useReducer } from "react";

const Context = createContext();

const ContextProvider = ({ children }) => {
  const initialState = {
    name: "",
    email: "",
    submittedData: [],
  };

  const reducer = (state, action) => {
    switch (action.type) {
      case "UPDATE_FIELD":
        return { ...state, [action.field]: action.value };
      case "RESET":
        return { ...state, name: "", email: "" };
      case "ADD":
        return {
          ...state,
          submittedData: [...state.submittedData, action.formData],
        };
      default:
        return state;
    }
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>
  );
};

export const Contexthook = () => {
  return useContext(Context);
};

export default ContextProvider;
