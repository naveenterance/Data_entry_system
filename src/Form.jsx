import React, { useReducer, createContext, useContext } from "react";

const FormContext = createContext();

const initialState = {
  name: "",
  email: "",
  submittedData: [],
};

const formReducer = (state, action) => {
  switch (action.type) {
    case "UPDATE_FIELD":
      return { ...state, [action.field]: action.value };
    case "RESET":
      return { ...state, name: "", email: "" };
    case "ADD":
      return {
        ...state,
        submittedData: [...state.submittedData, action.formData],
      };
    default:
      return state;
  }
};

const FormProvider = ({ children }) => {
  const [state, dispatch] = useReducer(formReducer, initialState);

  return (
    <FormContext.Provider value={{ state, dispatch }}>
      {children}
    </FormContext.Provider>
  );
};

const useFormContext = () => {
  return useContext(FormContext);
};

const Form = (props) => {
  const { state, dispatch } = useFormContext();

  const handleInputChange = (e) => {
    const { name, value } = e.target;
    dispatch({ type: "UPDATE_FIELD", field: name, value });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const formData = { name: state.name, email: state.email };
    dispatch({ type: "ADD", formData });
    dispatch({ type: "RESET" });

    props.setEntries([...state.submittedData, formData]);
  };

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="name"
          value={state.name}
          onChange={handleInputChange}
          placeholder="name"
        />

        <input
          type="email"
          name="email"
          value={state.email}
          onChange={handleInputChange}
          placeholder="email"
        />

        <button type="submit">Submit</button>
      </form>
    </div>
  );
};

export { FormProvider, useFormContext, Form };
