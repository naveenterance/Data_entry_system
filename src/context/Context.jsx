import React, { createContext, useContext, useReducer } from "react";

export const Context = createContext();

export const ContextProvider = ({ children }) => {
  const initialState = {
    firstName: "",
    lastName: "",
    email: "",
    phone: "",
    country: "",
    age: "",
    street: "",
    town: "",
    postcode: "",
    visits: "0",
    entries: [],
  };

  const reducer = (state, action) => {
    switch (action.type) {
      case "UPDATE_FIELD":
        return { ...state, [action.field]: action.value };
      case "RESET":
        return {
          ...state,
          firstName: "",
          lastName: "",
          email: "",
          phone: "",
          country: "",
          age: "",
          street: "",
          town: "",
          postcode: "",
          visits: "0",
        };
      case "ADD":
        return {
          ...state,
          entries: [...state.entries, action.formData],
        };
      default:
        return state;
    }
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>
  );
};

export const Contexthook = () => {
  return useContext(Context);
};
