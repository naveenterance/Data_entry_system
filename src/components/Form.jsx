import React from "react";
import { Contexthook } from "../context/Context";

const Form = () => {
  const { state, dispatch } = Contexthook();

  const handleInputChange = (e) => {
    const { name, value } = e.target;

    if (name == "phone") {
      if (/^\d{0,15}$/.test(value)) {
        dispatch({ type: "UPDATE_FIELD", field: name, value });
      }
    } else if (name == "age") {
      if (/^\d{0,2}$/.test(value)) {
        dispatch({ type: "UPDATE_FIELD", field: name, value });
      }
    } else {
      dispatch({ type: "UPDATE_FIELD", field: name, value });
    }
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const formData = {
      firstName: state.firstName,
      lastName: state.lastName,
      email: state.email,
      phone: state.phone,
      country: state.country,
      age: state.age,
      street: state.street,
      town: state.town,
      postcode: state.postcode,
      visits: state.visits,
    };
    dispatch({ type: "ADD", formData });
    dispatch({ type: "RESET" });
  };
  const incrementVisits = () => {
    const visitCount = parseInt(document.getElementById("visits").value, 10);

    dispatch({ type: "UPDATE_FIELD", field: "visits", value: visitCount + 1 });
  };
  const decrementVisits = () => {
    const visitCount = parseInt(document.getElementById("visits").value, 10);

    visitCount > 0 &&
      dispatch({
        type: "UPDATE_FIELD",
        field: "visits",
        value: visitCount - 1,
      });
  };

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="firstName"
          value={state.firstName}
          onChange={handleInputChange}
          placeholder="First Name"
        />
        <input
          type="text"
          name="lastName"
          value={state.lastName}
          onChange={handleInputChange}
          placeholder="Last Name"
        />

        <input
          type="email"
          name="email"
          value={state.email}
          onChange={handleInputChange}
          placeholder="email"
        />
        <input
          type="text"
          name="phone"
          value={state.phone}
          onChange={handleInputChange}
          placeholder="phone"
        />
        <select name="country" id="country" onChange={handleInputChange}>
          <option value="India">India</option>
          <option value="Australia">Australia</option>
          <option value="Europa">Europa</option>
          <option value="japan">japan</option>
        </select>
        <input
          type="text"
          name="age"
          value={state.age}
          onChange={handleInputChange}
          placeholder="age"
        />
        <input
          type="text"
          name="street"
          value={state.street}
          onChange={handleInputChange}
          placeholder="street"
        />
        <input
          type="text"
          name="town"
          value={state.town}
          onChange={handleInputChange}
          placeholder="town"
        />
        <input
          type="text"
          name="postcode"
          value={state.postcode}
          onChange={handleInputChange}
          placeholder="postcode"
        />

        <input
          id="visits"
          type="text"
          name="visits"
          value={state.visits}
          readOnly
        />
        <div onClick={incrementVisits}>Increment</div>
        <div onClick={decrementVisits}>decrement</div>
        <button type="submit">Submit</button>
      </form>
    </div>
  );
};

export default Form;
