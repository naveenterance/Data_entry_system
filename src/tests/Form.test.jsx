import React from "react";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Contexthook } from "../context/Context";
import Form from "../components/Form";
import renderer from "react-test-renderer";

jest.mock("../context/Context", () => ({
  Contexthook: jest.fn(),
}));

describe("Form component", () => {
  it("renders without crashing ", () => {
    Contexthook.mockReturnValue({ state: {}, dispatch: jest.fn() });

    const { container } = render(<Form />);

    expect(container).toMatchSnapshot();
  });

  it("updates form state on input change", () => {
    const dispatchMock = jest.fn();
    Contexthook.mockReturnValue({ state: {}, dispatch: dispatchMock });

    const { getByPlaceholderText } = render(<Form />);

    fireEvent.change(getByPlaceholderText("First Name"), {
      target: { value: "John" },
    });

    fireEvent.change(getByPlaceholderText("Last Name"), {
      target: { value: "Brennan" },
    });

    fireEvent.change(getByPlaceholderText("email"), {
      target: { value: "jb@gmail.com" },
    });

    expect(dispatchMock).toHaveBeenCalledWith({
      type: "UPDATE_FIELD",
      field: "firstName",
      value: "John",
    });

    expect(dispatchMock).toHaveBeenCalledWith({
      type: "UPDATE_FIELD",
      field: "lastName",
      value: "Brennan",
    });

    expect(dispatchMock).toHaveBeenCalledWith({
      type: "UPDATE_FIELD",
      field: "email",
      value: "jb@gmail.com",
    });
  });

  it("submits form data on button click", () => {
    const dispatchMock = jest.fn();
    Contexthook.mockReturnValue({
      state: { firstName: "John", lastName: "Brennan", email: "jb@gmail.com" },
      dispatch: dispatchMock,
    });

    const { getByText } = render(<Form />);

    fireEvent.click(getByText("Submit"));

    expect(dispatchMock).toHaveBeenCalledWith({
      type: "ADD",
      formData: {
        firstName: "John",
        lastName: "Brennan",
        email: "jb@gmail.com",
      },
    });

    expect(dispatchMock).toHaveBeenCalledWith({
      type: "RESET",
    });
  });
});
