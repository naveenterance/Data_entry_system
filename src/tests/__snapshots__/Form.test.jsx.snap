// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders without crashing  1`] = `
<div>
  <div>
    <form>
      <input
        name="firstName"
        placeholder="First Name"
        type="text"
        value=""
      />
      <input
        name="lastName"
        placeholder="Last Name"
        type="text"
        value=""
      />
      <input
        name="email"
        placeholder="email"
        type="email"
        value=""
      />
      <input
        name="phone"
        placeholder="phone"
        type="text"
        value=""
      />
      <select
        id="country"
        name="country"
      >
        <option
          value="India"
        >
          India
        </option>
        <option
          value="Australia"
        >
          Australia
        </option>
        <option
          value="Europa"
        >
          Europa
        </option>
        <option
          value="japan"
        >
          japan
        </option>
      </select>
      <input
        name="age"
        placeholder="age"
        type="text"
        value=""
      />
      <input
        name="street"
        placeholder="street"
        type="text"
        value=""
      />
      <input
        name="town"
        placeholder="town"
        type="text"
        value=""
      />
      <input
        name="postcode"
        placeholder="postcode"
        type="text"
        value=""
      />
      <input
        id="visits"
        name="visits"
        readonly=""
        type="text"
        value=""
      />
      <div>
        Increment
      </div>
      <div>
        decrement
      </div>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
